README
BBS-Client version 0.01
=======================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2007 by Cornelius

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension BBS::Client.

0.02  2007 Sep 20 12:56:37 AM
	- have added some feature like 

0.01  Tue Sep 11 00:28:24 2007
	- original version; created by h2xs 1.23 with options
		-AX BBS::Client



Makefile.PL
use 5.008008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    AUTHOR          => 'Cornelius Lin (cornelius.howl@gmail.com)',
    ABSTRACT        => 'BBS Client',
    dist            => {
        COMPRESS    => 'gzip',
        SUFFIX      => '.gz',
    },
    NAME              => 'BBS::Client',
    VERSION_FROM      => 'lib/BBS/Client.pm', # finds $VERSION
    PREREQ_PM         => { 
				Encode => 0 ,
				Time::HiRes => 0 ,
				Net::Telnet => 0 ,
				Text::Iconv => 0 ,
				Lingua::ZH::Wrap => 0 ,
		}, 
    EXE_FILES       => [ <script/*> ],
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/BBS/Client.pm', # retrieve abstract from module
       AUTHOR         => 'Cornelius <cornelius.howl@gmail.com>') : ()),
);


lib/BBS/Client/Scheme.pm
package BBS::Client::Scheme;
use strict;
use utf8;
our $esc = chr(27);

# -----------------------
# Bs2
# -----------------------
our %bs2_scheme = (
	'userid' 			=> qr/您的帳號/,
	'passwd' 			=> qr/您的密碼/,
	'press_any_key' 	=> qr/請按.*?任意鍵.*?繼續/,
	'repeat_login' 		=> qr/您想踢掉其他重複的 login.*?嗎？/,
	'hotboards' 		=> qr/熱門看板排行榜/,
	'main_menu' 		=> qr/主功能表/,
	'wrong_userid' 		=> qr/錯誤的使用者代號/,
	'wrong_passwd' 		=> qr/密碼輸入錯誤/,

# Articles
	'article_ask_label' => qr/發表文章於.*?看板/,
	'article_list' 		=> qr/\[1\;33\;46m\s*文章列表/,
	'article_no_content'=> qr/此文章無內容/,
	'article_posted'	=> qr/順利貼出文章/,
	'browse_bar' 		=> qr/瀏覽 P\.(\d+)/,
	'browse_bar_finish' => qr/文章選讀/,

# Userlist
	'userlist_bar' 			=> qr{46m 網友列表},
	'userlist_board_friend' => qr{\d+\s+$esc\[36m(.+?)\s+(.+?)\s+?},
);

our %bs2_cmd_scheme = (
	'next_page' => ' ',
	'search_board' => 's',
	'quit' => 'q',

# Articles
	'article_post' => "\x10",
	'article_menu' => "\x18",
	'article_save' => "S\n",

# Userlist 
	'userlist_show' =>  "\025\t4\n"  ,

);

# -----------------------
# Sayya
# -----------------------
our %sayya_scheme = (
	'userid' 				=> qr/請輸入代號/,
	'passwd' 				=> qr/請輸入密碼/,
	'press_any_key' 		=> qr/請按任意鍵繼續/,
	'repeat_login' 			=> qr/您想刪除其他重複的 login \(Y\/N\)嗎/,
	'hotboards' 			=> qr/熱門看板排行榜/,
	'main_menu' 			=> qr/主功能表/,
	'wrong_userid' 			=> qr/錯誤的使用者代號/,
	'wrong_passwd' 			=> qr/密碼輸入錯誤/,
	'article_list' 			=> qr/看板《.*?》/,
	'article_no_content' 	=> qr/此文章無內容/,
	'browse_bar' 			=> qr/瀏覽 P\.(\d+.*?\d+\%).*/,
	'browse_bar_finish' 	=> qr/文章選讀.*/,
	'edit_menu' 			=> qr/存檔選項/,
	'posted'				=> qr/順利貼出文章/,
	'editing'				=> qr/編輯文章/,
);

our %sayya_cmd_scheme = (
	'next_page' => ' ',
	'search_board' => "s\n",
	'quit' => 'q',
);

# ---------------
# ptt
# ---------------
our %ptt_scheme = (
	'userid' => qr/請輸入代號/,
	'passwd' => qr/請輸入您的密碼/,
	'press_any_key' => qr/請按.*?任意鍵.*?繼續/,
	'repeat_login' => qr/您想刪除其他重複的 login.*?嗎？/,
	'hotboards' => qr/熱門看板排行榜/,
	'main_menu' => qr/主功能表/,
	'wrong_userid' => qr/錯誤的使用者代號/,
	'wrong_passwd' => qr/密碼輸入錯誤/,
	'article_list' => qr/文章選讀/,
	'article_no_content' => qr/此文章無內容/,
	'browse_bar' => qr/.*瀏覽 第 (\d+\/\d+) 頁.*?\(\s{1,2}\d{1,2}\%\).*/,
	'browse_bar2' => qr/.*瀏覽 第 \d+\/\d+ 頁.*?此頁有控制碼.*/,
	'browse_bar_finish' => qr/.*瀏覽 第 \d+\/\d+ 頁 \(100\%\).*/,
);

our %ptt_cmd_scheme = (
	'next_page' => ' ',
	'search_board' => 's',
	'quit' => 'q',
);



1;



lib/BBS/Client.pm
package BBS::Client;
use 5.008008;
use strict;
use warnings;
use utf8;
require Exporter;
our @ISA = qw(Exporter);

use Switch 'Perl6';
use Encode;
use Time::HiRes qw(usleep);
use Net::Telnet;
use BBS::Client::Scheme;
use BBS::Client::Userlist;

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use BBS::Client ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw( ) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw( );
our $VERSION = '0.02';

BEGIN {

	}

my $esc = chr(27);
my $buffer_delay = 1*10**5;
my $buffer_timeout = 10;
$|=1;
binmode STDOUT , ":utf8";

sub new 
{
	my ($class,$ego) = @_;
	bless $ego , $class;
	$ego->scheme();
	$ego->init();
	return $ego;
}


sub scheme {
	my $ego = shift;
	# my $scheme_name = shift || 'bs2';
	given( $ego->{sys} ) {
		when 'bs2' {
			$ego->{prompt} = \%BBS::Client::Scheme::bs2_scheme; 
			$ego->{cmd} = \%BBS::Client::Scheme::bs2_cmd_scheme; 
		}
		when 'ptt' {
			$ego->{prompt} = \%BBS::Client::Scheme::ptt_scheme; 
			$ego->{cmd} = \%BBS::Client::Scheme::ptt_cmd_scheme;
		}
		when 'sayya' {
			$ego->{prompt} = \%BBS::Client::Scheme::sayya_scheme; 
			$ego->{cmd} = \%BBS::Client::Scheme::sayya_cmd_scheme; 
		}
		default {
			$ego->{prompt} = \%BBS::Client::Scheme::bs2_scheme; 
			$ego->{cmd} = \%BBS::Client::Scheme::bs2_cmd_scheme; 
		}
	}
}

# private method declaration
sub sendkey 
{
	my $ego = shift;
	my $key = shift;
	$ego->{t}->put( $key );
}

sub dump_buffer 
{
	my $ego = shift;
	my $tmp = shift;
	$tmp =~ s/\r//g;
	$tmp =~ s/$esc\[/\*\[/g;
	print "\n=============================\n";
	print decode('big5',$tmp);
	print "\n=============================\n";
}


sub dump
{
	my $ego = shift;
	my $tmp = $ego->{t}->get();
	$tmp =~ s/\r//g;
	$tmp =~ s/$esc\[/\*\[/g;
	print "\n=============================\n";
	print decode('big5',$tmp);
	print "\n=============================\n";
}


sub getscreen 
{
	my $ego = shift;
	usleep( $buffer_delay );
	my $str = $ego->{t}->get( Timeout => $buffer_timeout );
	print $str = decode('Big5',$str);
	return $str;
}

sub init
{
	my $ego = shift;
	$ego->{t} = Net::Telnet->new(
			Port => 23,
			Timeout => 30,
			Errmode => \&error
			);
	$ego->{t}->open( $ego->{host} );
}

sub prepare_dir
{
	my $ego = shift;
	mkdir( $ENV{HOME} . "/BBS" );
	mkdir( $ENV{HOME} . "/BBS/$ego->{host}/") unless(-d "./$ego->{host}");
	mkdir( $ENV{HOME} . "/BBS/$ego->{host}/$ego->{board}") unless(-d "./$ego->{host}/$ego->{board}");
}

sub fetch_article_content {
	my $ego = shift;
	my $buf = '';
	my $out = '';
	my $p = 1;
	my $l = 1;
	my $esc_jump_ptn = qr{$esc\[(\d+);1H};

	my $buf2='';
	my $last_endline = 0;
	my $do_cut = 0;
	while( my $str = $ego->{t}->get() ) {
		$buf2 .= $str;
		my $buf_de = decode('Big5',$buf2);
		if( $buf_de =~ $ego->{prompt}{browse_bar} ) {
			print "($1)\r";
			$ego->sendkey( $ego->{cmd}{next_page} );
			$buf .= $buf2;
			$buf2 ='';
		} elsif( $buf_de =~ $ego->{prompt}{browse_bar_finish} ) {
			$buf .= $buf2;
			$buf2 ='';
			$buf = decode('Big5',$buf);
			last;
		} elsif( $buf_de =~ $ego->{prompt}{article_no_content} ) {
			print "no content        \n";
			$ego->sendkey(" j\n");
			return ' ';
		}
	} 

	my @lines = split( /\n/ ,$buf);
	for my $index ( 0 .. $#lines ) 
	{
		my $line = $lines[ $index ];
		# find control char
		while ( $line =~ $esc_jump_ptn ) {
			my $blankline = '';
			for( $l+1 .. $1 ) { $blankline .= "\n" }
			$l = $1;
			$line =~ s/$esc_jump_ptn/$blankline/;
		}

		if ( $line eq '' ) {
			--$l;
		} else {
			$line =~ s/$ego->{prompt}{'browse_bar'}//g;
			$line =~ s/$ego->{prompt}{'browse_bar_finish'}//g;
			++$l;
			$out .= $line . "\n";
		}
	}
	$out =~ s/($esc\[K
				|\r
				|$esc\[(\d*?;?)*m
				|^$esc\[;H$esc\[2J
			)//gx;
	return $out;
}


# public method declaration
sub login 
{
	my $ego = shift;
	my $user = shift;
	my $pass = shift;
	my $buf = '';
	while( my $str = $ego->{t}->get() ) {

		my $buf_de = decode('Big5' , $buf .= $str );

		if ( $buf_de =~ $ego->{prompt}{userid} ) {
			$ego->sendkey( $user."\n");
			$buf = '';
		} 
		elsif( $buf_de =~ $ego->{prompt}{passwd} ) {
			$ego->sendkey( $pass."\n");
			$buf = '';
		} 
		elsif(  $buf_de =~ $ego->{prompt}{repeat_login}  ) {
			$ego->sendkey( "n\n" );
			$buf = '';
		} 
		elsif( $buf_de =~ $ego->{prompt}{press_any_key} ) {
			$ego->sendkey( $ego->{cmd}{quit} );
			$buf = '';
		} 
		elsif( $buf_de =~ $ego->{prompt}{hotboards} ) {
			$ego->sendkey( $ego->{cmd}{quit});
			$buf = '';
		} 
		elsif( $buf_de =~ $ego->{prompt}{main_menu} ) {
			print "get main menu\n";
			$buf = '';
			return 1;
		} 
		elsif( $buf_de =~ $ego->{prompt}{wrong_userid} ) {
			print "wrong userid\n";
			$buf = '';
			return 0;
		} 
		elsif( $buf_de =~ $ego->{prompt}{wrong_passwd} ) {
			print "wrong password\n";
			$buf = '';
			return 0;
		} 
		elsif( $buf_de =~ m/您有.+?篇文章尚未完成/ ) {
			$ego->sendkey("q\n"); # forget it
			$buf = '';
		}
	}
}

sub enter_board 
{
	my $ego = shift;
	my $board = shift;
	$ego->{board}=$board;
	$ego->sendkey( $ego->{cmd}{search_board} . $board . "\n");
	my $buf = '';
	while( my $str = $ego->{t}->get() ) {
		$buf .= $str;
		my $buf_de = decode('big5' , $buf );
		if ( $buf_de =~ $ego->{prompt}{press_any_key} ) {
			$ego->sendkey( $ego->{cmd}{quit} );
			$buf = '';
		}  elsif( $buf_de =~ $ego->{prompt}{article_list} )  {
			$buf = '';
			return 1;
		}
	}
}

sub fetch_articles 
{
	my ( $ego , $start , $end ) = @_;
	$ego->prepare_dir();
	$ego->sendkey( "$start\n\n" );
	for my $index ( $start .. $end ) {

		print "      fetching item $index...\r";
		my $c = $ego->fetch_article_content();
		$ego->sendkey( $ego->{cmd}{next_page} ); 

		open FH , sprintf("> ./%s/%s/%d.txt" , $ego->{host} , $ego->{board} , $index );
		binmode FH , ":utf8";
		print FH $c;
		close FH;
		print "item $index saved.            \n";

	}
	print "done. :)\n";
	$ego->sendkey( $ego->{cmd}{quit} );
}

sub offline 
{
	my $ego = shift;
	$ego->{t}->close();
}


# 
# User list
sub enter_userlist
{
	my $ego = shift;
	$ego->sendkey( $ego->{cmd}->{userlist_show} );  # <Ctrl-U> <Tab> <4> <Enter>
}


sub read_screen
{
	my ( $ego , $pattern ) = @_; 		# pattern to stop read
	my $buf = '';
	while( my $str = $ego->{t}->get() ) { 		# fetch screen
		$buf .= $str;
		my $buf_de = decode('Big5',$buf );
		return $buf_de if( $buf_de =~ m{$pattern} );
	}
}

sub userlist_write_log 
{
	my $ego = shift;
	my $action = shift;
	my %user = @_;
	my $log_filename = $ENV{HOME} . "/$ego->{host}/$ego->{board}.log";
	open FH , ">>" , $log_filename ;
	binmode FH , ":utf8";
	print FH "$action : $user{ID} - $user{NICK} \n";
	close FH;
}

sub listen_userlist 
{
	my $ego = shift;
	my %all_ids;
	my %cur_ids;
	my %last_ids;
	my $timestamp;
	my $level = 0;
	my %pad_ids;

	while(1) 
	{
		# send update key
		usleep(3*10**5);
		$ego->sendkey("s");

		my $screen = $ego->read_screen( $ego->{prompt}{userlist_bar} );

		# add id to current id list 
		while( $screen =~ m{$ego->{prompt}{userlist_board_friend}}g )
		{ 
			my ( $gid , $gnick )=( $1, $2);
			if ( ! exists( $all_ids{$gid} ) ) {
				$all_ids{$gid} = 1;
				$ego->userlist_write_log( 
					'new guest' , 
					( 'ID' => $gid , 'NICK' => $gnick )
				);
				print "new guest found \@ $ego->{board} : $gid ( $gnick ) \n";
			}
			$cur_ids{$gid}=1;
		} 


		# update timestamp at first time	
		$timestamp = time() if ( ! $timestamp );
		if( time() - $timestamp > 2 )	# update current id list to last id list ( every 2 sec )
		{
			$timestamp = time();
			use POSIX qw(strftime);
			my $now_string = strftime "%a %b %e %H:%M:%S %Y", localtime;

			# compare with last id list
			foreach my $id ( keys %cur_ids )  # find new join visitors and existed visitors
			{ 
				unless( $last_ids{ $id } ) {
					# new join
					$pad_ids{ $id } = $level;
					for( 0 .. $pad_ids{ $id } ) { print "---|" ; }
					print "  $id enter at $now_string  " . $all_ids{$id} . " times. \n";
					$all_ids{$id}++;
					$level++;
				}
			}

			foreach my $id ( keys %last_ids ) 
			{
				unless ( $cur_ids{ $id } ) { 
					for( 0 .. $pad_ids{ $id } ) { print "   |" ; }
					print "- $id leave at: $now_string \n";
					$level--;
				}
			}
			
			# update last id list
			%last_ids = %cur_ids;
			foreach my $h ( keys %cur_ids ) { delete $cur_ids{$h}; }

		}
	}
}

sub wait_for
{
	my $ego = shift;
	my $pattern = shift;
	my $buf = '';
	while( my $str = $ego->{t}->get() ) {
		$buf .= $str;
		my $buf2 = decode('Big5',$buf);
		if( $buf2 =~ m{$pattern} ) {
			return 1;
		}
	}
}

sub conv_utf8_to_big5
{
	my ($ego,$c) = @_;

	# check if content is utf8 encoding
	if( utf8::is_utf8( $c ) {
		# convert encoding from utf8 to big5
		# because BBS sytem use big5 encoding
		my $enc = Text::Iconv->new("utf8", "big5");
		my $content_big5 = $enc->convert( $c );
		return if( $enc->retval );
		return $content_big5;
	} else {
		return $c;
	}
}


sub post_article
{
	my ($ego,$title,$content) = @_;
	
	# send post command	
	$ego->sendkey(  $ego->{cmd}{article_post} );

	# make sure that we are going to post text
	return unless ( $ego->wait_for( $ego->{prompt}{article_ask_label} ) );

	# skip label
	$ego->sendkey("\n");
	
	# send title
	usleep( 2 * 10 ** 5 );
	$ego->sendkey( $title . "\n" );
	
	# signature
	$ego->sendkey("\n");
	usleep( 2 * 10 ** 5 );
	$ego->sendkey($content);
	
	usleep( 5 * 10 ** 5 );
	$ego->sendkey(  $ego->{cmd}{article_menu} );
	
	usleep( 2 * 10 ** 5 );
	$ego->sendkey(  $ego->{cmd}{article_save} );
	
	usleep( 2 * 10 ** 5 );
	$ego->sendkey( $ego->{cmd}{next_page} );
	#$ego->dump();
	#return unless ( $ego->wait_for( $ego->{prompt}{article_posted} ) );

	1;
}

sub error 
{
	print "error\n";
	exit;
}
1;
__END__

# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

BBS::Client - A Client Module For BBS Systems

=head1 SYNOPSIS

	use BBS::Client;
	my $o = BBS::Client->new( {
			host => 'bs2.to',
			sys =>  'bs2' ,  # system type ( ptt or maple )
			});

	if( $o->login($user,$pass) ) {

		$o->enter_board($board);
		$o->fetch_articles( $start , $end );

	} else {

		print "failed..\n";

	}

=head1 DESCRIPTION

To connect with BBS systems to backup articles , I wrote this
module , which is implemented with Net::Telnet.  You can fetch
articles from BBS Systems , such like ptt.cc , ptt2.cc and bs2.to

This module works on different BBS Systems by given specific scheme
to emit corresponding behaviors. You can also add your scheme to 
this module. ( BBS::Client::Scheme )

=head1 SEE ALSO

Net::Telnet

=head1 AUTHOR

Cornelius, E<lt>cornelius.howl@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Cornelius

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.
